nifi-nar-bundles/nifi-graph-bundle/nifi-other-graph-services/src/main/java/org/apache/nifi/graph/TinkerpopClientService.java

```diff
@@ -0,0 +1,525 @@
+            String username = context.getProperty(USER_NAME).evaluateAttributeExpressions().getValue();
+            String password = context.getProperty(PASSWORD).getValue();
+            builder.credentials(username, password);
+        }
+
+        Cluster cluster = builder.create();
+
+        transitUrl = String.format("gremlin%s://%s:%s%s", usesSSL ? "+ssl" : "",
+                String.join(",", hosts), cluster.getPort(), cluster.getPath());
+
+        return cluster;
+    }
+
+    protected Map<String, String> scriptSubmission(String query, Map<String, Object> parameters, GraphQueryResultCallback handler) {
+        try {
+            Client client = cluster.connect();
+            Iterator<Result> iterator = client.submit(query, parameters).iterator();
+            long count = 0;
+            while (iterator.hasNext()) {
+                Result result = iterator.next();
+                Object obj = result.getObject();
+                if (obj instanceof Map) {
+                    handler.process((Map)obj, iterator.hasNext());
+                } else {
+                    handler.process(new HashMap<>() {{
+                        put("result", obj);
+                    }}, iterator.hasNext());
+                }
+                count++;
+            }
+
+            Map<String, String> resultAttributes = new HashMap<>();
+            resultAttributes.put(NODES_CREATED, NOT_SUPPORTED);
+            resultAttributes.put(RELATIONS_CREATED, NOT_SUPPORTED);
+            resultAttributes.put(LABELS_ADDED, NOT_SUPPORTED);
+            resultAttributes.put(NODES_DELETED, NOT_SUPPORTED);
+            resultAttributes.put(RELATIONS_DELETED, NOT_SUPPORTED);
+            resultAttributes.put(PROPERTIES_SET, NOT_SUPPORTED);
+            resultAttributes.put(ROWS_RETURNED, String.valueOf(count));
+
+            return resultAttributes;
+
+        } catch (Exception ex) {
+            throw new ProcessException(ex);
+        }
+    }
+
+    protected Map<String, String> bytecodeSubmission(String s, Map<String, Object> map, GraphQueryResultCallback graphQueryResultCallback) {
+        String hash = DigestUtils.sha256Hex(s);
+        Script compiled;
+
+        if (this.traversalSource == null) {
+            this.traversalSource = createTraversal();
+        }
+        int rowsReturned = 0;
+
+        if (compiledCode.containsKey(hash)) {
+            compiled = compiledCode.get(hash);
+        } else {
+            compiled = groovyShell.parse(s);
+            compiledCode.put(s, compiled);
+        }
+
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug(map.toString());
+        }
+
+        Binding bindings = new Binding();
+        map.forEach(bindings::setProperty);
+        bindings.setProperty("g", traversalSource);
+        bindings.setProperty("log", getLogger());
+        try {
+            Object result = groovyShell.parse(s, bindings).run();
```

Comment:
You don't appear to be using the pre-compiled code to do the submission. Am I missing something in how `groovyShell` works here?

Reply:
@MikeThomsen  -- good catch, I updated it to use the compiled code.